core/src/main/java/org/lflang/federated/serialization/FedCustomPythonSerialization.java

```diff
@@ -0,0 +1,165 @@
+    return serializedVarName + ".buf";
+  }
+
+  private String initializeCustomSerializer() {
+    return "if (self->custom_serializer == NULL) {\n"
+        + "PyObject* pName = PyUnicode_DecodeFSDefault(\"%s\");\n"
+            .formatted(customSerializerPackage)
+        + "PyObject* pModule = PyImport_Import(pName);\n"
+        + "Py_DECREF(pName);\n"
+        + "if (PyErr_Occurred()) PyErr_Print();\n"
+        + "if (pModule == NULL) lf_print_error_and_exit(\"Could not load the custom serializer module '%s'.\");\n"
+            .formatted(customSerializerPackage)
+        + "PyObject* SerializerClass = PyObject_GetAttrString(pModule, \"Serializer\");\n"
+        + "if (SerializerClass == NULL) lf_print_error_and_exit(\"Could not find class 'Serializer' in module '%s'.\");\n"
+            .formatted(customSerializerPackage)
+        + "Py_DECREF(pModule);\n"
+        + "self->custom_serializer = PyObject_CallObject(SerializerClass, NULL);\n"
+        + "Py_DECREF(SerializerClass);\n"
+        + "if (self->custom_serializer == NULL) lf_print_error_and_exit(\"Could not instantiate class 'Serializer' in module '%s'.\");\n"
+            .formatted(customSerializerPackage)
+        + "lf_print_log(\"Loaded custom serializer module %s.\\n\");\n"
+            .formatted(customSerializerPackage)
+        + "}\n";
+  }
+
+  @Override
+  public StringBuilder generateNetworkSerializerCode(String varName, String originalType) {
+    StringBuilder serializerCode = new StringBuilder();
+    // Check that self->custom_serializer is not null
+    serializerCode.append(this.initializeCustomSerializer());
+    // Define the serialized PyObject
+    serializerCode
+        .append(
+            "PyObject *serializer_serialize = PyObject_GetAttrString(self->custom_serializer,"
+                + " \"serialize\");\n"
+                + "PyObject *args = PyTuple_Pack(1, ")
+        .append(varName)
+        .append(");\n")
+        .append(
+            "PyObject *serialized_pyobject = PyObject_CallObject(serializer_serialize, args);\n")
+        .append("if (serialized_pyobject == NULL) {\n")
+        .append("    if (PyErr_Occurred()) PyErr_Print();\n")
+        .append("    lf_print_error_and_exit(\"Could not serialize object.\");\n")
+        .append("}\n")
+        .append("Py_buffer ")
+        .append(serializedVarName)
+        .append(";\n")
+        .append("int returnValue = PyBytes_AsStringAndSize(serialized_pyobject, (char**)&")
+        .append(serializedVarName)
+        .append(".buf, &")
+        .append(serializedVarName)
+        .append(".len);\n");
+    serializerCode.append("Py_XDECREF(serializer_serialize);\n");
+    serializerCode.append("Py_XDECREF(args);\n");
+    // Error check
+    serializerCode.append("if (returnValue == -1) {\n");
+    serializerCode.append("    if (PyErr_Occurred()) PyErr_Print();\n");
+    serializerCode.append(
+        "    lf_print_error_and_exit(\"Could not serialize " + serializedVarName + ".\");\n");
```

Comment:
Again, this looks like an opportunity to define a utility C function.  The code generator could then just a generate a call to that function.

Reply:
Yes this is a great idea. It simplifies the code a lot and increases readability.